### ISSUE TITLE
Module Routing as Children

### ISSUE THREAD

**Issue by MonasteryJohn** [angular/angular#10647]
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report
[X] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

If I define a module fully, with it's own routes and components and then apply that module as a child of a module intended to be the parent module, the routes still act independently.

Example

AdminModule
     Admin Component (has router-outlet)
     SetupComponent
     UserModule 
         UserComponent (has router-outlet)
         NewUserComponent

Setup URL is /admin/setup and outputs in the admin component router-outlet (good) 
New User URL is /user/new-user and it outputs in the User components router-outlet which ends up the main app-component router-outlet instead of the admin component. (bad)

**Expected/desired behavior**
In the example above, the desire would be that my route would be /admin/user/new-user and that the user component would load in the router-outlet of the  admin component, rather than on it's own.

Basically, I want to be able to navigate to components of child modules as though they were defined as child routes of the parent modules routing configuration.  Prior to moving my app to modules, I would just stack route definitions and import the child routes (with children of their own) in the higher level route configs. I realize I could expand my definition of modules, making the user components part of the admin module, but for separation I'd rather not do that.

**Reproduction of the problem**
If the current behavior is a bug or you can illustrate your feature request better with an example, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (you can use this template as a starting point: http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5).

**What is the expected behavior?**
I'm not sure what the expected behavior is, I'm assuming that the way it acts is correct according to the current design of the routing and module systems.

**What is the motivation / use case for changing the behavior?**

My application is intended to be a composition of multiple modules. Each of those modules is intended to be able to stand on it's own, as they are re-used in other sibling applications in different ways. That said, once composed, the routes and outlets should make reasonably good sense to both the end users and developers. I shouldn't have to change the definitions of my modules to get them to stack appropriately.

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.5 (Router 3.0.0-RC.1)
 
* **Browser:** [Chrome XX | Firefox XX | IE XX ] 
 
* **Language:** [TypeScript 1.8.10]

**Comment by brandonroberts** [comment]
Take a look at the router developer guide. Its doing what you are looking for. There is a root module, and two submodules with their own routes. Those submodules are imported into the root module and the routes are merged together.

tl;dr

The main module uses `RouterModule.forRoot(routes)`
The sub modules use `RouterModule.forChild(routes)`

https://angular.io/docs/ts/latest/guide/router.html

**Comment by MonasteryJohn** [comment]
@brandonroberts thanks for that, but that is exactly what I've done and the outcome is what I stated above. The route /admin/users/new-user is invalid but the route /users/new-user works fine. I could set my route in the users routing to be prefixed with admin but this doesn't resolve the issue of getting it into the correct router outlet. 

const AdminRoutes: Routes = [
    {path: "admin", component: AdminComponent}
] 

export const adminRouting = RouterModule.forChild(AdminRoutes)

...and...

const UserRoutes: Routes = [
       path: "users",
       component: userComponent,
       children: [
           {path: "new-user", component: newUserComponent}
       ]
] 

export const userRouting = RouterModule.forChild(UserRoutes)

The one level nesting of children off of root, works fine. What I need to nest additional children after the first level and that doesn't seem to be supported right now.

**Comment by brandonroberts** [comment]
I see. Importing submodules with routes into the main module only flattens the configurations together. The `users` routes aren't a child of `admin` unless you specify them that way in the route configuration.

**Comment by MonasteryJohn** [comment]
Changing the route to start with admin will fix the path issue, but not the outlet issue. I could flatten my admin module by removing all of the child modules and putting all of their components into the admin module, but that doesn't feel quite right to me. A lot of my modules are re-used in other parallel applications and prior to the module change, that worked very well. I like the concept of the modules, if only this part of it worked. 

I'll work around it if I have to, but I figured I'd try and get an official answer on support for child of child routing.

**Comment by hypdeb** [comment]
I have the same issue. Everything gets loaded into the router outlet of the component bootstraped in the root module.

**Comment by MonasteryJohn** [comment]
I found a way to work around this for now. Basically, I am defining my routes the way that I used to, but this time at the top child level. For example my admin route:

    const AdminRoutes: Routes = [
       {
          path: 'admin',
          component: AdminComponent,
          children: [
              ...setupRoutes
          ]
       }
    ]

    export const adminRouting = RouterModule.forChild(AdminRoutes)

My setup routes file is imported from a sub area, which defines routes of it's own, including more children. The catch is that this file exports the "Routes" object and not the RouterModule.forChild result. 

After that is setup, I removed the child and sub-child routes from the submodule definitions. I then had to export all of the components used in the routes, from each of the submodules. Once I did that, the routes started working just like I wanted them to.

I'm not a big fan of this workaround but at least now I can take advantage of the rest of the RC5 features. Preferably the module definitions themselves being imported into parents would be enough to layer the route definitions. I guess that's a decision the angular team can make. Hopefully this helps others.

**Comment by Tragetaschen** [comment]
Router configuration in a feature module is transported via a module import (wait, what?). When you look at the source code of the `forChild` method, the actual routes are put in the `providers` and become part of dependency injection. Every non-lazy-loaded module's providers are hoisted to the root `AppModule` and lose their hierarchical information. From that root module, the routes are then injected into the router. The flattening is essentially the reason, why any sub-router-outlet is ignored.

The two "working" options are

1) Your workaround basically circumvents the entire advantages feature modules are supposed to bring.
2) Use a lazy loaded module and `loadChildren`. That works, because then there is a new injector to provide the routes and the Router can see the existing component tree and router-outlet. Lazy loading shouldn't be a requirement for that.

A possible solution could be to change `loadChildren` to accept an actual module class instead of the string. The router should then behave the same as with lazy loaded modules, but instead load it eagerly.



There was a great blog article lately. I resisted to mark questionable parts in the analysis with [a pepper 🌶](https://medium.com/@mbostock/what-makes-software-good-943557f8a488).

**Comment by brandonroberts** [comment]
Take a look at the conversation on #10958. With RC6 you can use `loadChildren` synchronously

**Comment by Tragetaschen** [comment]
Yeah, I was going to open something like that :)

**Comment by DzmitryShylovich** [comment]
Duplicate of https://github.com/angular/angular/issues/10958?

**Comment by mp3por** [comment]
I am sure that I do not fully understand Angular2 so I have created a 'play' project in order to help me and others understand how to create route hierarchies and properly handle unmatched routes. I have done things following the latest tutorial on routing and feature modules (https://angular.io/docs/ts/latest/guide/ngmodule.html). Here is the link: https://github.com/mp3por/Angular2FeatureModulesRouting

Please review it, lets bring it to life and help everyone :)

**Comment by gudh** [comment]
@MonasteryJohn this works for me, thanks!

**Comment by DzmitryShylovich** [comment]
@MonasteryJohn is this still valid?

**Comment by mirjalolbahodirov** [comment]
@MonasteryJohn Thank you very much! You saved me a lot of time! It works like a charm!

**Comment by StefanRein** [comment]
Like @Tragetaschen said, defining the Sub-Feature Modules in the Routing worked really well.
For someone searching like me, here is an example:

**app.module.ts**
```
@NgModule({
    imports: [
        AppRoutingModule
    ]
})
export class AppModule {}
```

**app-routing.module.ts**
```
@NgModule({
    imports: [
        RouterModule.forRoot([
            {
                path: '',
                redirectTo: '/registration',
                pathMatch: 'full'
            },
            {
                path: 'registration',
                loadChildren: () => RegistrationModule
            },
            {
                path: 'dashboard',
                loadChildren: () => DashboardModule
            },
            {
                path: '**',
                component: PageNotFoundComponent
            }
        ]),
    ],
    exports: [RouterModule]
})
export class AppRoutingModule {}
```

**dashboard.module.ts**
```
@NgModule({
    imports: [
        DashboardRoutingModule
    ]
})
export class DashboardModule {}
```

**dashboard-routing.module.ts**
```
@NgModule({
    imports: [
        RouterModule.forChild([
            {
                path: '',
                component: DashboardComponent,
                children: [
                    {
                        path: 'overview',
                        loadChildren: () => DashboardOverviewModule
                    },
                    {
                        path: '',
                        redirectTo: 'overview',
                        pathMatch: 'full'
                    }
                ]
            }
        ])
    ],
    exports: [RouterModule]
})
export class DashboardRoutingModule {}
```

**dashboard-overview.module.ts**
```
@NgModule({
    imports: [
        DashboardOverviewRoutingModule
    ]
})
export class DashboardOverviewModule {}
```

**dashboard-overview-routing.module.ts**
```
@NgModule({
    imports: [
        RouterModule.forChild([
            {
                path: '',
                component: DashboardOverviewComponent
            }
        ])
    ],
    exports: [RouterModule]
})
export class DashboardOverviewRoutingModule {}
```

This scales really well and I can define the used route names in the parent module, which is great and eliminates collisions.
Imagine the route is defined in the feature module and I am importing two:
FooOverview
BarOverview
and for some reason the route path would be in both 'overview' => conflict
Now I can define the paths in the parent module.

**Beware:** Do not import the module again in the parent module. It is enough to import it in the RouterModule (with loadchildren).

**Comment by MikeDub** [comment]
@StefanRein, I tried implementing your example and logically it seems that it would work, however I am getting this error below, any ideas why? 

> ERROR in Error encountered resolving symbol values statically. Function calls are not supported. Consider replacing the function or lambda with a reference to an exported function (position 15:
> 19 in the original .ts file), resolving symbol appRoutes in ...../src/app/app-routing.module.ts, resolving symbol AppRoutingModule in ..../src/app/app-routing.module.ts

**Comment by bennyt2** [comment]
@StefanRein's answer worked for me in development (`ng serve`), but I got the same error as @MikeDub when compiling to production (`ng serve --prod`). The syntax in the [current Routing documentation](https://angular.io/docs/ts/latest/guide/router.html#!#lazy-loading-route-config) worked for me in both development and production. Here's the relevant snippet:

> Give it a loadChildren property (not a children property!), set to the address of the AdminModule. The address is the AdminModule file location (relative to the app root), followed by a # separator, followed by the name of the exported module class, AdminModule.

```
{
  path: 'admin',
  loadChildren: 'app/admin/admin.module#AdminModule',
},
```

**Comment by StefanRein** [comment]
@MikeDub @bennyt2 

Which TypeScript and Angular version are you both using?
The difference I see is, that I am using the predefined Webpack and co preferences from [https://www.npmjs.com/package/@ionic/app-scripts](https://www.npmjs.com/package/@ionic/app-scripts)

Running the "equally" ionic serve production command compiles fine.
Did one of you try the recommendation of the message?

Like:

```
export function loadAdminModule() {
    return AdminModule;
}
```

and then inserting the exported loadAdminModule function. 
Does this generate any different error message?


An excerpt of my package.json with the versions
```
{
  "dependencies": {
    "@angular/common": "4.1.3",
    "@angular/compiler": "4.1.3",
    "@angular/compiler-cli": "4.1.3",
    "@angular/core": "4.1.3",
    "@angular/forms": "4.1.3",
    "@angular/http": "4.1.3",
    "@angular/platform-browser": "4.1.3",
    "@angular/platform-browser-dynamic": "4.1.3",
    "@angular/router": "4.1.3",
    "@ionic-native/core": "3.4.2",
    "ionic-angular": "3.2.1",
    "rxjs": "5.1.1",
    "sw-toolbox": "3.4.0",
    "zone.js": "^0.8.9"
  },
  "devDependencies": {
    "@angular/language-service": "^4.1.3",
    "@ionic/app-scripts": "1.3.7",
    "@types/node": "^7.0.21",
    "html-webpack-plugin": "^2.28.0",
    "typescript": "~2.2.1",
    "webpack-merge": "^4.1.0",
    "cross-env": "~5.0.0"
  }
}
```

**Comment by bennyt2** [comment]
@StefanRein Below is an excerpt of my packages. I used angular-cli to build the project (I used angular-cli to start a project yesterday morning, so I haven't added anything to the build process; it's just the base angular-cli build process with `ng serve` and `ng serve --prod`).

```
{
  "dependencies": {
    "@angular/common": "^4.0.0",
    "@angular/compiler": "^4.0.0",
    "@angular/core": "^4.0.0",
    "@angular/forms": "^4.0.0",
    "@angular/http": "^4.0.0",
    "@angular/platform-browser": "^4.0.0",
    "@angular/platform-browser-dynamic": "^4.0.0",
    "@angular/router": "^4.0.0",
    "angularfire2": "^4.0.0-rc.0",
    "core-js": "^2.4.1",
    "firebase": "^3.9.0",
    "rxjs": "^5.1.0",
    "zone.js": "^0.8.4"
  },
  "devDependencies": {
    "@angular/cli": "1.0.6",
    "@angular/compiler-cli": "^4.0.0",
    "@types/jasmine": "2.5.38",
    "@types/node": "~6.0.60",
    "codelyzer": "~2.0.0",
    "jasmine-core": "~2.5.2",
    "jasmine-spec-reporter": "~3.2.0",
    "karma": "~1.4.1",
    "karma-chrome-launcher": "~2.1.1",
    "karma-cli": "~1.0.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "karma-coverage-istanbul-reporter": "^0.2.0",
    "protractor": "~5.1.0",
    "ts-node": "~2.0.0",
    "tslint": "~4.5.0",
    "typescript": "~2.2.0"
  }
}
```

I did try the recommendation of the message, like so:

```
export function loadEventModule() {
    return EventModule;
}
```
Granted it's with my code, not the admin/dashboard sample code in your example, but the reason it doesn't work is clear: The LoadChildrenCallback has this type signature: `(): Type<any> | NgModuleFactory<any> | Promise<Type<any>> | Observable<Type<any>>`. Simply returning a class (in my case, `EventModule`), doesn't meet that criteria.

```
ERROR in /home/ben/Documents/projects/tmio-angular/src/app/meet/meet-routing.module.ts (15,35): Argument of type '({ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetD...' is not assignable to parameter of type 'Route[]'.
  Type '{ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetDe...' is not assignable to type 'Route'.
    Type '{ path: string; component: typeof MeetDetailComponent; children: { path: string; loadChildren: ty...' is not assignable to type 'Route'.
      Types of property 'children' are incompatible.
        Type '{ path: string; loadChildren: typeof EventModule; }[]' is not assignable to type 'Route[]'.
          Type '{ path: string; loadChildren: typeof EventModule; }' is not assignable to type 'Route'.
            Types of property 'loadChildren' are incompatible.
              Type 'typeof EventModule' is not assignable to type 'LoadChildren'.
                Type 'typeof EventModule' is not assignable to type 'LoadChildrenCallback'.
                  Type 'typeof EventModule' provides no match for the signature '(): Type<any> | NgModuleFactory<any> | Promise<Type<any>> | Observable<Type<any>>'
```
I also tried wrapping it in a promise, but that didn't work either. Here's that error trace:

```
ERROR in /home/ben/Documents/projects/tmio-angular/src/app/meet/meet-routing.module.ts (15,35): Argument of type '({ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetD...' is not assignable to parameter of type 'Route[]'.
  Type '{ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetDe...' is not assignable to type 'Route'.
    Type '{ path: string; component: typeof MeetDetailComponent; children: { path: string; loadChildren: Pr...' is not assignable to type 'Route'.
      Types of property 'children' are incompatible.
        Type '{ path: string; loadChildren: Promise<typeof EventModule>; }[]' is not assignable to type 'Route[]'.
          Type '{ path: string; loadChildren: Promise<typeof EventModule>; }' is not assignable to type 'Route'.
            Types of property 'loadChildren' are incompatible.
              Type 'Promise<typeof EventModule>' is not assignable to type 'LoadChildren'.
                Type 'Promise<typeof EventModule>' is not assignable to type 'LoadChildrenCallback'.
                  Type 'Promise<typeof EventModule>' provides no match for the signature '(): Type<any> | NgModuleFactory<any> | Promise<Type<any>> | Observable<Type<any>>'
```

**Comment by bennyt2** [comment]
@StefanRein Below is an excerpt of my packages. I used angular-cli to build the project (I used angular-cli to start a project yesterday morning, so I haven't added anything to the build process; it's just the base angular-cli build process with `ng serve` and `ng serve --prod`).

```
{
  "dependencies": {
    "@angular/common": "^4.0.0",
    "@angular/compiler": "^4.0.0",
    "@angular/core": "^4.0.0",
    "@angular/forms": "^4.0.0",
    "@angular/http": "^4.0.0",
    "@angular/platform-browser": "^4.0.0",
    "@angular/platform-browser-dynamic": "^4.0.0",
    "@angular/router": "^4.0.0",
    "angularfire2": "^4.0.0-rc.0",
    "core-js": "^2.4.1",
    "firebase": "^3.9.0",
    "rxjs": "^5.1.0",
    "zone.js": "^0.8.4"
  },
  "devDependencies": {
    "@angular/cli": "1.0.6",
    "@angular/compiler-cli": "^4.0.0",
    "@types/jasmine": "2.5.38",
    "@types/node": "~6.0.60",
    "codelyzer": "~2.0.0",
    "jasmine-core": "~2.5.2",
    "jasmine-spec-reporter": "~3.2.0",
    "karma": "~1.4.1",
    "karma-chrome-launcher": "~2.1.1",
    "karma-cli": "~1.0.1",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "karma-coverage-istanbul-reporter": "^0.2.0",
    "protractor": "~5.1.0",
    "ts-node": "~2.0.0",
    "tslint": "~4.5.0",
    "typescript": "~2.2.0"
  }
}
```

I did try the recommendation of the message, like so:

```
export function loadEventModule() {
    return EventModule;
}
```
Granted it's with my code, not the admin/dashboard sample code in your example, but the reason it doesn't work is clear: The LoadChildrenCallback has this type signature: `(): Type<any> | NgModuleFactory<any> | Promise<Type<any>> | Observable<Type<any>>`. Simply returning a class (in my case, `EventModule`), doesn't meet that criteria.

```
ERROR in /home/ben/Documents/projects/tmio-angular/src/app/meet/meet-routing.module.ts (15,35): Argument of type '({ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetD...' is not assignable to parameter of type 'Route[]'.
  Type '{ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetDe...' is not assignable to type 'Route'.
    Type '{ path: string; component: typeof MeetDetailComponent; children: { path: string; loadChildren: ty...' is not assignable to type 'Route'.
      Types of property 'children' are incompatible.
        Type '{ path: string; loadChildren: typeof EventModule; }[]' is not assignable to type 'Route[]'.
          Type '{ path: string; loadChildren: typeof EventModule; }' is not assignable to type 'Route'.
            Types of property 'loadChildren' are incompatible.
              Type 'typeof EventModule' is not assignable to type 'LoadChildren'.
                Type 'typeof EventModule' is not assignable to type 'LoadChildrenCallback'.
                  Type 'typeof EventModule' provides no match for the signature '(): Type<any> | NgModuleFactory<any> | Promise<Type<any>> | Observable<Type<any>>'
```
I also tried wrapping it in a promise, but that didn't work either. Here's that error trace:

```
ERROR in /home/ben/Documents/projects/tmio-angular/src/app/meet/meet-routing.module.ts (15,35): Argument of type '({ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetD...' is not assignable to parameter of type 'Route[]'.
  Type '{ path: string; component: typeof MeetListComponent; } | { path: string; component: typeof MeetDe...' is not assignable to type 'Route'.
    Type '{ path: string; component: typeof MeetDetailComponent; children: { path: string; loadChildren: Pr...' is not assignable to type 'Route'.
      Types of property 'children' are incompatible.
        Type '{ path: string; loadChildren: Promise<typeof EventModule>; }[]' is not assignable to type 'Route[]'.
          Type '{ path: string; loadChildren: Promise<typeof EventModule>; }' is not assignable to type 'Route'.
            Types of property 'loadChildren' are incompatible.
              Type 'Promise<typeof EventModule>' is not assignable to type 'LoadChildren'.
                Type 'Promise<typeof EventModule>' is not assignable to type 'LoadChildrenCallback'.
                  Type 'Promise<typeof EventModule>' provides no match for the signature '(): Type<any> | NgModuleFactory<any> | Promise<Type<any>> | Observable<Type<any>>'
```

**Comment by StefanRein** [comment]
@bennyt2 Please show me the EventModule
Did you provide the Decorator? Because then the return value of `NgModuleFactory<any>` should be fine?!

Like 

```javascript
const COMPONENTS = [];

@NgModule({
    imports: [],
    declarations: [
        ...COMPONENTS
    ],
    exports: [
        ...COMPONENTS
    ]
})
export class EventModule {
}
```

**Comment by bennyt2** [comment]
@StefanRein Yes, the decorator is there. Here is all of my relevant code

app-module.ts
```
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HttpModule } from '@angular/http';

import { AppComponent } from './app.component';
import {CoreModule} from "./core/core.module";
import {SharedModule} from "./shared/shared.module";
import {RouterModule, Routes} from "@angular/router";
import {MeetModule} from "./meet/meet.module";
import {EventModule} from "./event/event.module";

const appRoutes: Routes = [
  { path: '',   redirectTo: '/', pathMatch: 'full' },
];

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    CoreModule,
    SharedModule,
    EventModule,
    MeetModule,
    RouterModule.forRoot(appRoutes)
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

meet-module.ts
```
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { MeetRoutingModule } from './meet-routing.module';
import { MeetDetailComponent } from './meet-detail/meet-detail.component';
import { MeetListComponent } from './meet-list/meet-list.component';
import {MeetService} from "./meet.service";

@NgModule({
  imports: [
    CommonModule,
    MeetRoutingModule
  ],
  providers: [
    MeetService
  ],
  declarations: [MeetDetailComponent, MeetListComponent],
})
export class MeetModule { }
```

meet-routing-module.ts
```
import {NgModule} from '@angular/core';
import {RouterModule} from '@angular/router';
import {MeetListComponent} from "./meet-list/meet-list.component";
import {MeetDetailComponent} from "./meet-detail/meet-detail.component";

@NgModule({
  imports: [
    RouterModule.forChild([
    {
      path: 'meets',
      component: MeetListComponent
    },
    {
      path: 'meets/:id',
      component: MeetDetailComponent,
      children: [
        {
          path: 'events',
          loadChildren: 'app/event/event.module#EventModule'
        }
      ]
    }
  ])],
  exports: [RouterModule]
})
export class MeetRoutingModule {
}
```

event-module.ts
```
import {NgModule, Type} from '@angular/core';
import { CommonModule } from '@angular/common';
import { EventListComponent } from './event-list/event-list.component';
import { EventService } from './event.service';
import {EventRoutingModule} from "./event-routing.module";

@NgModule({
  imports: [
    CommonModule,
    EventRoutingModule
  ],
  declarations: [EventListComponent],
  providers: [EventService]
})
export class EventModule { }
```

event-routing-module.ts
```
import {NgModule} from '@angular/core';
import {Routes, RouterModule} from '@angular/router';
import {EventListComponent} from "./event-list/event-list.component";

const routes: Routes = [
  {
    path: '',
    component: EventListComponent
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class EventRoutingModule {
}
```

**Comment by StefanRein** [comment]
Your problem goes away with this:

```javascript
export function exportMeetModule() {
    return MeetModule;
}

@NgModule({
    imports: [
        RouterModule.forRoot([
            {
                path: '',
                redirectTo: 'meet',
                pathMatch: 'full'
            },
            {
                path: 'meet',
                loadChildren: exportMeetModule
            }
        ], {
            useHash: true
        })
    ],
    exports: [
        RouterModule
    ]
})
export class AppRoutingModule {
}

```

But then I am getting another error:

```
ERROR Error: Uncaught (in promise): Error: Runtime compiler is not loaded
Error: Runtime compiler is not loaded
```

Here is the repository on github:
https://github.com/StefanRein/angular-cli-modules-subroutes

**Comment by shairez** [comment]
What we ended up doing is just exporting an array of routes and point it to the children property.

It achieves the same goal, but without a hack

So if we need to lazy load, we use `forChild` and `loadChildren`
But if we just want routes to be placed on top of former routes, we just do

```
{path: 'someChild', children: SOME_CHILDREN_ROUTES }
```

And we get the same result

**Comment by VigneshSAN** [comment]
I am having error like 

> Cannot match any routes. URL Segment: 'users/users-list'
    at

app.routing.ts

`import { NgModule } from "@angular/core";
import { Routes, RouterModule } from "@angular/router";
import { DashboardComponent } from "./dashboard/dashboard.component";
import { UsersModule } from "./users/users.module";
import { MerchantsModule } from "./merchants/merchants.module";
import { LogisticsModule } from "./logistics/logistics.module";
import { ShopsModule } from "./shops/shops.module";
import { OrdersModule } from "./orders/orders.module";

const routes: Routes = [
    {
        path: '',
        redirectTo: '/dashboard',
        pathMatch: 'full',
    },
    {
        path: 'dashboard',
        component: DashboardComponent,
        data: {
            title: 'Dashboard'
        },
        children: [
            {
                path: 'users',
                loadChildren: () => UsersModule
            },
            {
                path: 'merchants',
                loadChildren: () => MerchantsModule
            },
            {
                path: 'logistics',
                loadChildren: () => LogisticsModule
            },
            {
                path: 'shops',
                loadChildren: () => ShopsModule
            },
            {
                path: 'orders',
                loadChildren: () => OrdersModule
            },
        ]
    }
];

@NgModule({
    imports: [RouterModule.forRoot(routes)],
    exports: [RouterModule]
})

export class AppRoutingModule { }`

And my Users.Module.ts

`import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { AddUsersComponent } from './add-users/add-users.component';
import { UsersListComponent } from './users-list/users-list.component';

const routes: Routes = [
  {
    path: '',
    data: {
      title: 'Users'
    },
    children: [
      {
        path: 'add-users',
        component: AddUsersComponent,
        data: {
          title: 'Add Users'
        }
      },
      {
        path: 'users-list',
        component: UsersListComponent,
        data: {
          title: 'Users List'
        }
      }
    ]
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class UsersRoutingModule { }
`

Kindly help me to solve this issues.

**Comment by adessilly** [comment]
thx a lot @StefanRein, it seems to be pretty elegant, but I haven't found any documentation about the way you load children : 

> loadChildren: () => DashboardOverviewModule

Is it standard or is it a trick you found by yourself ?

I'm used to write loadChildren this way : 

> loadChildren: '[path]/[moduleFileName]#[module]'

**Comment by adessilly** [comment]
Maybe the only difference I see is that your way is not lazy loading because you return a hard reference to the module ?!

**Comment by atscott** [comment]
Closing as obsolete / working as intended. Routes are hierarchical and must be defined as such. In order to have a set of routes be children of another, they must be defined in the `children`/`loadChildren` property. Child routes are much easier to share now as well after https://github.com/angular/angular/pull/45700
